Add nether biomes API

jaskarth opened the pull request:
Add a new class and mixin to allow modders to inject their own biomes into the nether in a compatible manner.

jaskarth committed:
Add nether biomes API

jaskarth committed:
Update fabric-biomes-v1.mixins.json

valoeghese commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java:
```diff
@@ -0,0 +1,21 @@
+package net.fabricmc.fabric.api.biomes.v1;
```
licence

valoeghese commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java:
```diff
@@ -0,0 +1,23 @@
+package net.fabricmc.fabric.impl.biome;
```
licencE

valoeghese commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheNetherDimension.java:
```diff
@@ -0,0 +1,25 @@
+package net.fabricmc.fabric.mixin.biome;
```
lIcenCE

valoeghese commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheNetherDimension.java:
```diff
@@ -0,0 +1,25 @@
+import net.fabricmc.fabric.impl.biome.NetherBiomesImpl;
+
+@Mixin(TheNetherDimension.class)
+public class MixinTheNetherDimension {
```
good

valoeghese commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Sets internal use only! Stores data that is used by the Nether dimension mixin.
+ */
+public class NetherBiomesImpl {
```
nice

valoeghese commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java:
```diff
@@ -0,0 +1,21 @@
+	 *
+	 * @param biome The biome to add.
+	 */
+	public static void addNetherBiome(Biome biome) {
```
approve

jaskarth committed:
fixed license

Earthcomputer commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheNetherDimension.java:
```diff
@@ -0,0 +1,41 @@
+
+@Mixin(TheNetherDimension.class)
+public class MixinTheNetherDimension {
+	@Redirect(method = "createChunkGenerator", at = @At(value = "INVOKE", target = "net/minecraft/class_4767.method_24404(Ljava/util/Set;)Lnet/minecraft/class_4767;"))
```
Wouldn't this be better as a `@ModifyArg`?

jaskarth replied:
Done

jaskarth committed:
smarter injection point

coderbot16 commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java:
```diff
@@ -0,0 +1,37 @@
+import net.fabricmc.fabric.impl.biome.NetherBiomesImpl;
+
+/**
+ * API that exposes internals' of Minecraft's nether biome code.
```
```suggestion
 * API that exposes the internals of Minecraft's nether biome code.
```

coderbot16 commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java:
```diff
@@ -0,0 +1,37 @@
+	/**
+	 * Adds a biome to the Nether generator. Biomes must set their own noise generation values in the biome settings class.
+	 *
+	 * @param biome The biome to add.
```
```suggestion
	 * @param biome The biome to add. Must not be null.
```

coderbot16 commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java:
```diff
@@ -0,0 +1,39 @@
+	private static final Set<Biome> netherBiomes = new HashSet<>();
+
+	public static void addNetherBiome(Biome biome) {
+		netherBiomes.add(biome);
```
Add an `Objects.requireNonNull` check here

coderbot16 commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheNetherDimension.java:
```diff
@@ -0,0 +1,41 @@
+public class MixinTheNetherDimension {
+	@ModifyArg(method = "createChunkGenerator", at = @At(value = "INVOKE", target = "Lnet/minecraft/class_4767;method_24404(Ljava/util/Set;)Lnet/minecraft/class_4767;"))
+	protected Set<Biome> modifyNetherBiomes(Set<Biome> set) {
+		Set<Biome> newSet = new HashSet<>(set); // the set in the parameter is immutable, so we construct our own
```
comments on the preceding line are more readable imo
```suggestion
		// the set in the parameter is immutable, so we construct our own
		Set<Biome> newSet = new HashSet<>(set);
```

coderbot16 commented on fabric-biomes-v1/src/main/resources/fabric-biomes-v1.mixins.json:
```diff
@@ -8,7 +8,8 @@
     "MixinSetBaseBiomesLayer",
-    "MixinVanillaLayeredBiomeSource"
+    "MixinVanillaLayeredBiomeSource",
+    "MixinTheNetherDimension"
```
Maintain alphabetical order please

coderbot16 commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java:
```diff
@@ -0,0 +1,39 @@
+	}
+
+	public static Set<Biome> getNetherBiomes() {
+		return netherBiomes;
```
Consider wrapping this with `Collections.unmodifiableSet`.

coderbot16 reviewed (changes requested):
Some easy fixes.

jaskarth committed:
Apply suggestions from coderbot's review

Co-Authored-By: coderbot16 <coderbot16@gmail.com>

jaskarth committed:
fix code style

jaskarth committed:
Merge branch '1.16' of https://github.com/SuperCoder7979/fabric into 1.16

liach commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheNetherDimension.java:
```diff
@@ -0,0 +1,42 @@
+import org.spongepowered.asm.mixin.injection.ModifyArg;
+import org.spongepowered.asm.mixin.injection.Redirect;
+
+import net.minecraft.class_4767;
```
Unused import

jaskarth committed:
remove redundant import

coderbot16 commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java:
```diff
@@ -0,0 +1,37 @@
+	private NetherBiomes() { }
+
+	/**
+	 * Adds a biome to the Nether generator. Biomes must set their own noise generation values in the biome settings class.
```
Perhaps elaborate on which settings and have a `{@link Biome.Settings}` as well

liach replied:
Need to detail that you need to set up the biome noise in settings properly or the biomes have literally zero chance to generate in the nether

jaskarth replied:
I think i will add some basic information and then write some detailed explanation on the wiki

coderbot16 reviewed (changes requested):
One final review, I promise! I think we should just clarify to API users what settings they need to tweak.

liach reviewed (approved):
Build passes. Style is good.

jaskarth committed:
improve docs and update yarn version

coderbot16 commented:
@Earthcomputer can we get a second review from you? Just want to clean up the reviews list :p

Juuxel commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java:
```diff
@@ -0,0 +1,38 @@
+	 * Adds a biome to the Nether generator. Biomes must set their own noise values in the MixedNoisePoint class for the biome to properly generate.
+	 *
+	 * @param biome The biome to add. Must not be null.
+	 * {@link Biome.MixedNoisePoint}
```
```suggestion
	 * @see Biome.MixedNoisePoint
```

coderbot16 commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java:
```diff
@@ -0,0 +1,38 @@
+	private NetherBiomes() { }
+
+	/**
+	 * Adds a biome to the Nether generator. Biomes must set their own noise values in the MixedNoisePoint class for the biome to properly generate.
```
If we're applying Juuz's comment,
```suggestion
	 * Adds a biome to the Nether generator. Biomes must set their own noise values in the {@link Biome.MixedNoisePoint} class for the biome to properly generate.
```

jaskarth committed:
Apply suggestions from code review

Co-Authored-By: coderbot16 <coderbot16@gmail.com>
Co-Authored-By: Juuxel <6596629+Juuxel@users.noreply.github.com>

coderbot16 commented:
I think this is ready for a final review from Player/modmuss. Can @Juuxel give this another look?

coderbot16 commented:
Almost forgot, we should probably increment the version number to 0.2.0. I'm not sure how this will interact with #491, or if we'll need to bump it to 0.3.0 in the other PR.

Prospector commented:
Bump to 0.2.0 in both and unless they're merged at the same time, bump the one that was merged last to 0.3.0

jaskarth committed:
bump version

dblsaiko commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java:
```diff
@@ -0,0 +1,38 @@
+/**
+ * API that exposes the internals of Minecraft's nether biome code.
+ */
+public class NetherBiomes {
```
Everywhere else in API (afaik) this class would be an interface with an INSTANCE field that NetherBiomesImpl implements (see e.g. net.fabricmc.fabric.api.client.model.ModelLoadingRegistry.) Why is this different here?

Not that it makes any real difference, but it would be more consistent.

jaskarth replied:
I wanted to be consistent with the rest of the biome API which follows the same pattern that I used.

SuperCoder7979 replied:
I wanted to be consistent with the rest of the biome API which follows the same pattern that I used.

Prospector replied:
Hm, yeah I didn't even notice we did the OverworldBiomes one any differently. Also I noticed in OverworldBiomes we had the class as `final`, maybe put that here too?

jaskarth replied:
fixed 👍

dblsaiko replied:
Yeah, if this style exists somewhere previously in API, then fair enough.

jaskarth committed:
mark classes as final

modmuss50 commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java:
```diff
@@ -0,0 +1,38 @@
+/**
+ * API that exposes the internals of Minecraft's nether biome code.
+ */
+public final class NetherBiomes {
```
Why not an interface like most of the other apis?

liach replied:
Should be fine as you can just change the method body as long as the method names and descriptors stay intact

modmuss50 commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java:
```diff
@@ -0,0 +1,42 @@
+public final class NetherBiomesImpl {
+	private NetherBiomesImpl() { }
+
+	private static final Set<Biome> netherBiomes = new HashSet<>();
```
Does this contain vanilla biomes?

modmuss50 replied:
I guess its not an issue, as its not exposed via the api, ingore me.

liach replied:
No, just extra biomes to be added

modmuss50 commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Sets for internal use only! Stores data that is used by the Nether dimension mixin.
+ */
+public final class NetherBiomesImpl {
```
Why make this a class and not add it to `InternalBiomeData`?

jaskarth replied:
That is a great question if i'm being honest

modmuss50 commented:
Module also needs to be marked to only run on 1.16+ versions

jaskarth committed:
Changes from modmuss's review

add 1.16 dependency, merge NetherBiomesImpl into InternalBiomeData, fix code style, and fix mixin target

modmuss50 reviewed (approved):
Implement #499 and then I think this is good to merge.

jaskarth committed:
Update build.gradle

jaskarth committed:
Update API to inform users that it is experimental

jaskarth committed:
remove beta annotion

jaskarth committed:
optimize imports
